master/pkg/tasks/task_gc.go

```diff
@@ -46,6 +46,9 @@ func (g GCCkptSpec) ToTaskSpec(allocationToken string) TaskSpec {
 	}
 	res.Environment = schemas.WithDefaults(env).(expconf.EnvironmentConfig)
 
+	resSpec := res.Environment.PodSpec()
```

Comment:
`resSpec` comes straight out of `schemas.WithDefaults(env).(expconf.EnvironmentConfig)` for which the default value given by the jsonschema is `null`, and `expconf.PodSpec.Merge` isn't nil-safe (the k8s code inside it does an immediate deref), so wouldn't this always panic?

Reply:
Local non-k8s run gets a panic exactly here
```
INFO[2021-09-27T09:15:48-04:00] starting checkpoint garbage collection        id=experiment-8-checkpoint-gc system=master type=checkpointGCTask
ERRO[2021-09-27T09:15:48-04:00] runtime error: invalid memory address or nil pointer dereference
goroutine 5042 [running]:
runtime/debug.Stack(0xc000eb3578, 0x34f27e0, 0x5162340)
        /usr/local/go/src/runtime/debug/stack.go:24 +0x9f
github.com/determined-ai/determined/master/pkg/actor.(*Ref).close(0xc000fa5500)
        /Users/bradleylaney/Code/determined/master/pkg/actor/ref.go:306 +0x9a5
panic(0x34f27e0, 0x5162340)
        /usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/determined-ai/determined/master/pkg/tasks.GCCkptSpec.ToTaskSpec(0x0, 0x0, 0x0, 0x0, 0xc00013a750, 0x24, 0xc00005c640, 0x36, 0x0, 0x0, ...)
        /Users/bradleylaney/Code/determined/master/pkg/tasks/task_gc.go:50 +0x333
github.com/determined-ai/determined/master/internal.(*checkpointGCTask).Receive(0xc0004f2c00, 0xc001070e40, 0x39919ef, 0x18)
        /Users/bradleylaney/Code/determined/master/internal/checkpoint_gc.go:55 +0x808
github.com/determined-ai/determined/master/pkg/actor.(*Ref).processMessage(0xc000fa5500, 0x34ed900)
        /Users/bradleylaney/Code/determined/master/pkg/actor/ref.go:253 +0x36e
github.com/determined-ai/determined/master/pkg/actor.(*Ref).run(0xc000fa5500)
        /Users/bradleylaney/Code/determined/master/pkg/actor/ref.go:265 +0xa5
created by github.com/determined-ai/determined/master/pkg/actor.newRef
        /Users/bradleylaney/Code/determined/master/pkg/actor/ref.go:91 +0x571  id=experiment-8-checkpoint-gc system=master type=checkpointGCTask
```

Reply:
I think we'd want to merge the other way, `g.LegacyConfig.PodSpec().Merge(resSpec)` since the receiver takes higher priority in the merge, and we prob want to just make merge nil-safe.

Reply:
> and expconf.PodSpec.Merge isn't nil-safe

There's a fix for this:

> Use schemas.Merge() and it will handle these nil cases automatically

Reply:
ahhhhhh ok I didn't properly read @rb-determined-ai's original instructions

Reply:
> schemas.Merge()

forgot this existed

Reply:
side note: this highlights we don't have a single test that makes sure GC runs OK, right? else it would've failed.

Reply:
https://determinedai.atlassian.net/browse/DET-6032

Reply:
We have an e2e test which should catch this.

Reply:
oh.. lol i see, it failed. i just totally missed it.

Reply:
> I think we'd want to merge the other way, g.LegacyConfig.PodSpec().Merge(resSpec) since the receiver takes higher priority in the merge

I think bradley's right that the order here isn't quite right.  Normally what we do for submitted experiments is:
- load user-provided values
- merge in values from templates
- merge in values from task container defaults
- call WithDefaults for any remaining values.

I think when we instantiate the `expconf.EnvironmentConfig{}` object above we should set the `g.LegacyConfig.PodSpec()` in there directly, since this is closest to the user-provided values, and therefore should have the highest priority (as the envvars do)